### PR TITLE
Sphinx gallery

### PR DESCRIPTION
**Pull request by tachella**
# Update

## Changes to Physics
- Added single pixel camera operator.
- Added single photon lidar operator.
- fixed blurFFT to be real SVD.
- small fixes to MRI.
- Decolorize has now SVD.

## Changes to demos
- Fixed and documented all existing demos.
- Added new sampling and basic demos.

## Changes to test
- Added a sampling test using a posterior which is available in closed form

## Changes to docs
- Improved organization of docs (welcome page)
- added docs for plotting utils
- sphinx gallery integration

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.md](../../CHANGELOG.md) (as applicable) (to come).

### CODE REVIEW

File: deepinv/tests/test_sampling.py [hunk 1]
```diff
@@ -17,84 +16,109 @@ def device():
...
+            sigma=1,
+        )
+    elif algo == "DDRM":
+        diff = dinv.sampling.DDRM(
+            denoiser=GaussianDenoiser(sigma_prior),
+            sigma_noise=sigma,
+            eta=1,
+            sigmas=np.linspace(1, 0, 100),
         )
+        out = dinv.sampling.DiffusionSampler(diff, clip=(-100, 100), max_iter=100)
```
Comment (matthieutrs) [inline deepinv/tests/test_sampling.py:57]: This test failed here in the last push with no apparent reason; maybe we can set a random seed to avoid having some random failuers?

File: deepinv/loss/regularisers.py [hunk 2]
```diff
@@ -87,7 +98,7 @@ def forward(self, y, x):
                 v.detach_()
                 u.detach_()
 
-        return z.view(-1).sqrt()
+        return self.weight * z.view(-1).sqrt()
```
Comment (matthieutrs) [inline deepinv/loss/regularisers.py:101]: I would not do this: this functions gives the spectral norm of the Jacobian, so if we multiply the max eigenvalue by `self.weight`, it is confusing. I would rather add the weight somewhere else. If we do not find anything clearer, then I would move the compute function inside a `def compute(self, ...)` method (or any other name) returning the proper computation, and the forward would just call `self.compute(...)` and multiply the output with `self.weight`. What do you think is best?

File: deepinv/optim/optimizers.py [hunk 3]
```diff
@@ -243,11 +243,13 @@ def update_params_fn(self, it):
 
     def init_params_fn(self):
         r"""
-        Get initialization parameters and re-initialized the parameters dictionary to its initial value. This is necessary if the parameters have been updated during optimizaiton, for example via backtracking.
+        Get initialization parameters and re-initialized the parameters' dictionary to its initial value.
```
Comment (matthieutrs) [inline deepinv/optim/optimizers.py:246]: Needs clarification, can assign myself to that.

File: examples/sampling/demo_ddrm.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+# We can now reconstruct a noisy measurement using the diffusion method.
+# We use the DDRM method from :class:`deepinv.sampling.DDRM`, which works with inverse problems that
+# have a closed form singular value decomposition of the forward operator.
+# The diffusion method requires a schedule of noise levels ``sigmas`` that are used to evaluate the denoiser.
+
+diff = dinv.sampling.DDRM(
+    denoiser=denoiser,
+    etab=1.0,
+    sigma_noise=sigma,
+    sigmas=np.linspace(1, 0, 100),  # uses 100 evaluations of the denoiser
```
Comment (matthieutrs) [inline examples/sampling/demo_ddrm.py:94]: Time issues on CPU

```suggestion
    sigmas=np.linspace(1, 0, 10),  # uses 10 evaluations of the denoiser; set to sigmas=np.linspace(1, 0, 10) for better results
```

**Review by matthieutrs — CHANGES_REQUESTED** [review]
Great PR! I'm requesting changes because I am afraid sphinx-gallery examples won't run with current parameters (e.g. number of iterations) that are too optimistic for CPU.

File: deepinv/tests/test_sampling.py (hunk 1 above)
Comment (tachella) [reply deepinv/tests/test_sampling.py:57]: Yes, good idea

File: deepinv/loss/regularisers.py (hunk 2 above)
Comment (tachella) [reply deepinv/loss/regularisers.py:101]: good point. Maybe we can add the weights somewhere in the train function, and avoid adding them to the losses one by one.

File: deepinv/physics/blur.py [hunk 5]
```diff
@@ -460,74 +467,92 @@ class BlurFFT(DecomposablePhysics):
...
+    plt.show()
+    plt.imshow(y.squeeze(0).permute(1, 2, 0).cpu().numpy())
+    plt.show()
+    plt.imshow(xhat.squeeze(0).permute(1, 2, 0).cpu().numpy())
+    plt.show()
+    plt.imshow(xhat2.squeeze(0).permute(1, 2, 0).cpu().numpy())
+    plt.show()
+
+    plt.imshow(physics.A(xhat).squeeze(0).permute(1, 2, 0).cpu().numpy())
+    plt.show()
```
[the remark this answers is not in the document]
Comment (tachella) [reply deepinv/physics/blur.py:558]: this code is just for debugging the blur operators. We can just comment it

File: examples/sampling/demo_sampling.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+#
+# Here we use the Unadjusted Langevin Algorithm (ULA) to sample from the posterior defined in
+# :class:`deepinv.sampling.ULA`.
+# The hyperparameter ``step_size`` controls the step size of the MCMC sampler,
+# ``regularization`` controls the strength of the prior and
+# ``iterations`` controls the number of iterations of the sampler.
+
+regularization = 0.9
+step_size = 0.01 * (sigma**2)
+iterations = int(5e3)
```
[the remark this answers is not in the document]
Comment (tachella) [reply examples/sampling/demo_sampling.py:111]: we can also reduce the image size

File: deepinv/__init__.py [hunk 7]
```diff
@@ -48,10 +48,14 @@
...
 __all__ += ["sampling"]
 
-from deepinv.loss import metric as metric
+from deepinv.loss import metric
 
 __all__ += ["metric"]
 
+__all__ += ["unfolded"]
+
+from deepinv import unfolded
```
Comment (tomMoral) [inline deepinv/__init__.py:57]: Reverse the two lines?
```suggestion
from deepinv import unfolded

__all__ += ["unfolded"]
```

File: docs/source/deepinv.loss.rst [hunk 8]
```diff
@@ -38,6 +38,18 @@ about the forward measurement process.
     deepinv.loss.SurePGLoss
     deepinv.loss.TVLoss
 
+Transforms
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (tomMoral) [inline docs/source/deepinv.loss.rst:42]:
```suggestion
Transforms
^^^^^^^^^^
```

File: docs/source/deepinv.loss.rst
Comment (tomMoral) [inline docs/source/deepinv.loss.rst:1]: Usually, this module level documentation is put as a docstring in the `__init__.py` file of the module, so here in `deepinv/loss/__init__.py`, so it can be accessed as

```python
import deepinv.loss
deepinv.loss.__doc__
```

It will also be generated with autodoc in the API description page.

File: docs/source/index.rst [hunk 9]
```diff
@@ -22,27 +34,35 @@ Here quick guide
...
+============================
+
+The preferred way to contribute to ``deepinv`` is to fork the `main
+repository <https://github.com/deepinv/deepinv/>`__ on GitHub,
+then submit a "Pull Request" (PR).
+
 
 Lead Developers
-==================
+============================
```
Comment (tomMoral) [inline docs/source/index.rst:61]: Usually, rst linters prefer that the length of the `====` is the same as the one of the text, for easier reading in text mode (in particular in docstrings).

File: examples/basics/demo_physics.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+#     :meth:`deepinv.physics.LinearPhysics.adjointness_test()`,
+#     and that it has a unit norm using :meth:`deepinv.physics.LinearPhysics.compute_norm()`.
+
+print(f"The linear operator has norm={physics.compute_norm(x):.2f}")
+
+if physics.adjointness_test(x) < 1e-5:
+    print("The linear operator has a well defined transpose")
+
+
+# %%
```
Comment (tomMoral) [inline examples/basics/demo_physics.py:88]: I think this should be a second example, maybe exposing simply more of the possibilities of the physic operator and having a specific decomposable operator example along side, maybe not in basics.

File: examples/basics/demo_physics.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.mask = 0.447
+        self.noise_model = dinv.physics.GaussianNoise(sigma=0.1)
+
+    def V_adjoint(self, x):
+        y = x[:, 0, :, :] * 0.6687 + x[:, 1, :, :] * 1.3132 + x[:, 2, :, :] * 0.2550
+        return y.unsqueeze(1)
```
Comment (tomMoral) [inline examples/basics/demo_physics.py:114]: Not sure this is the best example as here  it is very similar to the original one.
So if you don't know what this is, it might be complex to get it.

Maybe doing a transform in Fourier where everyone knows the SVD or something a little bit more didactic changing from RGB to another encoding  (with a matrix) and computing the SVD explicitly might give better insights.

File: examples/basics/demo_physics.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    Converts RGB images to grayscale.
+
+    Signals must be tensors with 3 colour (RGB) channels, i.e. [*,3,*,*]
+    The measurements are grayscale images.
+
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.mask = 0.447
```
Comment (tomMoral) [inline examples/basics/demo_physics.py:109]: seems unused? It is not clear what you need this so this should be explained in the example.

File: examples/basics/demo_physics.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+end = time.time()
+print(f"Elapsed time for LinearPhysics: {end - start:.2f} seconds")
+
+start = time.time()
+for i in range(10):
+    xlin2 = physics2.A_dagger(x)
+    xprox2 = physics2.prox_l2(x, y2, 0.1)
+
+end = time.time()
+print(f"Elapsed time for DecomposablePhysics: {end - start:.2f} seconds")
```
Comment (tomMoral) [inline examples/basics/demo_physics.py:165]: If this goes faster than `0.01`, I recommend using higher precision or `.2e`.

**Review by tomMoral — COMMENTED** [review]
A small round of comments but this looks pretty good!

File: examples/sampling/demo_ddrm.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    "Lionel-Messi-Argentina-2022-FIFA-World-Cup_%28cropped%29.jpg"
+)
+res = requests.get(url)
+x = imread(BytesIO(res.content)) / 255.0
+
+x = torch.tensor(x, device=dinv.device, dtype=torch.float).permute(2, 0, 1).unsqueeze(0)
+x = torch.nn.functional.interpolate(
+    x, scale_factor=0.5
+)  # reduce the image size for faster eval
+
```
[the remark this answers is not in the document]
Comment (matthieutrs) [reply examples/sampling/demo_ddrm.py:44]: Added in demo_sampling as well.

File: docs/source/conf.py
[the remark this answers is not in the document]
Comment (matthieutrs) [reply docs/source/conf.py:1]: Ok, thanks! Not updating for now, FYI @tachella

File: docs/source/deepinv.loss.rst
Comment (matthieutrs) [reply docs/source/deepinv.loss.rst:1]: Might be good indeed @tachella. Leaving as is for this merge but might be worth doing that in the future?

Comment (tachella) [reply docs/source/deepinv.loss.rst:1]: yes, definitely. thanks for the recommendation!

File: examples/basics/demo_physics.py (hunk 11 above)
Comment (matthieutrs) [reply examples/basics/demo_physics.py:114]: We should probably split this example in two and add a `Physics` section to the example.

**Review by matthieutrs — APPROVED** [review]
